### PR TITLE
Loading scan mode data for IN5, PANTHER, SHARP

### LINKED ISSUE #30780: LoadILLDiffraction segfault
### Expected behavior

Diffraction data from ILL instruments D1B, D2B, and D20 should be successfully loaded and further processed in Mantid.

### Actual behavior

There's a segmentation violation error that causes an immediate crash of the workbench when loading is attempted on data from the most recent (and ongoing as of 26.02.2021) data-taking cycle.

### Steps to reproduce the behavior

If you have access to ILL data: 
1. Try to load any data from the experiments or internalUse data for D1B, D2B, and D20.
2. Observe the subsequent segmentation violation.

If you don't have data to test, you can contact me.

### Platforms affected
all

### LINKED ISSUE #30781: Scan-mode data loading for IN5 and PANTHER
There is new data taking mode currently available for IN5 and PANTHER for scanning the sample properties, such as position (omega), temperature, etc. This workflow is similar to the one used in ILL Diffraction. However, there is currently no support for this data structure for these two instruments. 

### Platforms affected
all

### PR DESCRIPTION
**Pull request by darominski**
This PR implements the new data taking workflow recently added to IN5 and PANTHER instruments at the ILL. The scan data structure is the same as in scanning diffraction instruments at the ILL, thus the scan mode for IN5 and PANTHER is loaded through LoadILLDiffraction, with minor changes, to avoid code duplication.

This PR also fixes a recent issue (#30780) caused by scan data type change, that resulted in segmentation violations when an attempt was made to load data from the current data-taking cycle. This was caused by the type change from "ILL_detectors_data_scan" to "NXill_detectors_data_scan" .

The LoadILLTest has been filled with missing tests for D1B, and D7, as well as the new mode for IN5.

**To test:**

1. Get the new unit test data for IN5 scan mode: ILL/IN5/199857.nxs
2. Try to load data, check the instrument viewer, check if the sample logs are filled, e.g. wavelength

To test the solution of #30780 , if you access to ILL data:
1. Try to load internalUse data from the cycle 211.
2. There should be no issues, the sampleLogs should be properly filles.

Fixes #30781 and 
Fixes #30780 and 
Part of #30902

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gvardany** [comment]
This has got conflicts now, could you resolve them?

File: docs/source/algorithms/LoadILLTOF-v2.rst [hunk 1]
```diff
@@ -17,6 +17,8 @@ To date this algorithm only supports: IN4, IN5, IN6, PANTHER, and SHARP.
 By default, this algorithm loads the data indexed by channels. To convert to time-of-flight, use the ConvertToTOF option.
 
 This algorithm also supports diffraction mode. In this case, the unit of the output workspace will be wavelength instead of time-of-flight or channel.
+The support for the omega scan measurement mode for IN5 and PANTHER is **not** handled by this loader, but instead the data
```
Comment (gvardany) [inline docs/source/algorithms/LoadILLTOF-v2.rst:20]: and SHARP

File: Framework/DataHandling/test/LoadILLDiffractionTest.h [hunk 2]
```diff
@@ -493,6 +494,65 @@ class LoadILLDiffractionTest : public CxxTest::TestSuite {
     checkTimeFormat(outputWS);
   }
 
+  void test_IN5_omega_scan() {
```
Comment (gvardany) [inline Framework/DataHandling/test/LoadILLDiffractionTest.h:497]: Can you add tests also for the PANTHER and SHARP numors?

File: Framework/DataHandling/src/LoadILLDiffraction.cpp [hunk 3]
```diff
@@ -496,18 +499,26 @@ void LoadILLDiffraction::fillStaticInstrumentScan(const NXUInt &data, const NXDo
   const std::vector<double> axis = getAxis(scan);
   const std::vector<double> monitor = getMonitor(scan);
 
+  size_t monitorIndex = 0;
+  size_t startIndex = NUMBER_MONITORS;
+  if (m_instName == "IN5" || m_instName == "PANTHER" || m_instName == "SHARP") {
```
Comment (gvardany) [inline Framework/DataHandling/src/LoadILLDiffraction.cpp:504]: This condition is checked twice, maybe define a variable `isSpectrometer` and reuse it.

**Review by gvardany — COMMENTED** [review]
This looks mostly alright, just few minor issues.
I'll check the functionality shortly.
I see now the possibility to factor out LoadILLScan, but there is no one best way of splitting. Right now, while it is confusing that TOF instruments are loaded with LoadILLDiffraction, it is the optimal solution minimizing code duplication. We can think this over later on, but it shouldn't block this. One general issue is that the tube numbering convention is different on diffraction instruments as compared to the rest, including the spectrometers. Can you make sure you map the data to tube/pixel the right way? You can check how LoadILLTOF does it. I suspect 2 of dimensions might be swapped, or be in reverse order.

File: Framework/DataHandling/test/LoadILLDiffractionTest.h (hunk 2 above)
Comment (darominski) [reply Framework/DataHandling/test/LoadILLDiffractionTest.h:497]: Not a problem for PANTHER, but there was no SHARP scan mode data the last time I checked. Do you know whether it was used in the last cycle?

Comment (gvardany) [reply Framework/DataHandling/test/LoadILLDiffractionTest.h:497]: You just added in the test data. I think it's 000104.

Comment (darominski) [reply Framework/DataHandling/test/LoadILLDiffractionTest.h:499]: Oh yes, absolutely! It's exactly that one.

**Comment by darominski** [comment]
> This looks mostly alright, just few minor issues.
> I'll check the functionality shortly.
> I see now the possibility to factor out LoadILLScan, but there is no one best way of splitting. Right now, while it is confusing that TOF instruments are loaded with LoadILLDiffraction, it is the optimal solution minimizing code duplication. We can think this over later on, but it shouldn't block this. One general issue is that the tube numbering convention is different on diffraction instruments as compared to the rest, including the spectrometers. Can you make sure you map the data to tube/pixel the right way? You can check how LoadILLTOF does it. I suspect 2 of dimensions might be swapped, or be in reverse order.

I verified that the loading order is conformant between `LoadILLTOF2` and `LoadILLDiffraction`, where iteration goes over tubes and over pixels in tubes, so data vectors are defined the same way. Swapping the order in `LoadILLDiffraction` for loading scanned PANTHER data leads to severe and obvious distortions in the counts distribution, namely the shape of the beam stop and the location of low counts pixels: they are moved from being distributed horizontally on the edge of the instrument to vertically around the edges which physically does not make sense and does not conform with non-scan distributions loaded with `LoadILLTOF2`.

**Review by gvardany — APPROVED** [review]
Looks good, nice catch with the nexus class change for the scans.

**Pull request merged**